vowpalwabbit/core/include/vw/core/learner.h

```diff
@@ -297,386 +149,175 @@ class learner
   /// \param i This is the offset used for the weights in this call. If using
   /// multiple regressors/learners you can increment this value for each call.
   /// \returns The prediction calculated by this reduction be set on
-  /// example::pred. If <code>E</code> is ::multi_ex then the prediction is set
-  /// on the 0th item in the list.
-  inline void NO_SANITIZE_UNDEFINED predict(E& ec, size_t i = 0)
-  {
-    assert((is_multiline() && std::is_same<multi_ex, E>::value) ||
-        (!is_multiline() && std::is_same<example, E>::value));  // sanity check under debug compile
-    details::increment_offset(ec, increment, i);
-    debug_log_message(ec, "predict");
-    _learn_fd.predict_f(_learn_fd.data, *_learn_fd.base, (void*)&ec);
-    details::decrement_offset(ec, increment, i);
-  }
+  /// example::pred. If the polymorphic_ex is ::multi_ex then the prediction is
+  /// set on the 0th item in the list.
+  void predict(polymorphic_ex ec, size_t i = 0);
```

Comment:
Perhaps we should rename `polymorphic_ex` to `polymorphic_ex_ref` to make it clear it is essentially a reference wrapper

Reply:
Fair point but `polymorphic_ex` already a pretty long name, so I'm hesitant to make it even longer.

Also, multi_ex holds example pointers inside and not example objects.

Reply:
That's true. Perhaps add a doc comment so at least we're clear that this is a reference type